pennylane/qnodes/base.py

```diff
@@ -336,13 +374,49 @@ def _construct(self, args, kwargs):
                 inside of this method, the quantum function returns incorrect values or if
                 both continuous and discrete operations are specified in the same quantum circuit
         """
-        # pylint: disable=attribute-defined-outside-init, too-many-branches
+        # pylint: disable=attribute-defined-outside-init, too-many-branches, too-many-statements
+
+        # flatten the args, replace each argument with a Variable instance carrying a unique index and name
+        full_argspec = inspect.getfullargspec(self.func)
+
+        # args
+        variable_name_strings = []
+        for variable_name, variable_value in zip(full_argspec.args, args):
+            variable_name_strings.append(self._determine_variable_name(variable_value, variable_name))
 
-        # flatten the args, replace each argument with a Variable instance carrying a unique index
-        arg_vars = [Variable(idx) for idx, _ in enumerate(_flatten(args))]
-        self.num_variables = len(arg_vars)
+        # varargs
+        len_diff = len(args) - len(full_argspec.args)
+        if len_diff > 0:
+            for idx, variable_value in enumerate(args[-len_diff:]):
+                variable_name = "{}[{}]".format(full_argspec.varargs, idx)
+
+                variable_name_strings.append(self._determine_variable_name(variable_value, variable_name))
+
+        self.arg_vars = [Variable(idx, name) for idx, name in enumerate(_flatten(variable_name_strings))]
+        self.num_variables = len(self.arg_vars)
         # arrange the newly created Variables in the nested structure of args
-        arg_vars = unflatten(arg_vars, args)
+        self.arg_vars = unflatten(self.arg_vars, args)
+
+        # kwargs
+        variable_name_strings = {}
+        self.kwarg_vars = {}
+        for variable_name in full_argspec.kwonlyargs:
+            if variable_name in kwargs:
+                variable_value = kwargs[variable_name]
+            else:
+                variable_value = full_argspec.kwonlydefaults[variable_name]
+
+            if isinstance(variable_value, np.ndarray):
+                variable_name_string = np.empty_like(variable_value, dtype=object)
+
+                for index in np.ndindex(*variable_name_string.shape):
+                    variable_name_string[index] = "{}[{}]".format(variable_name, ",".join([str(i) for i in index]))
+
+                kwarg_variable = [Variable(idx, name=name, is_kwarg=True) for idx, name in enumerate(_flatten(variable_name_string))]
+            else:
+                kwarg_variable = Variable(0, name=variable_name, is_kwarg=True)
+
+            self.kwarg_vars[variable_name] = kwarg_variable
```

Comment:
This is great, but not so sure it should be placed directly in construct (since construct is a very central method to the QNode). I would suggest either a private method that can be called here, or even an auxiliary function in this file.

Reply:
It would also be good to have more commenting of the code above

Reply:
I refactored this part.